### PR TITLE
Referentially separate backoff functions

### PR DESCRIPTION
**Pull request by kriskowal**
Without this change, it’s possible for backoff strategies to share state through a common random number generator, which leads to data races.

This change changes a backoff strategy to a constructor for lockless but referentially independent backoff functions. The random number generator option is now a constructor function.

### CODE REVIEW

**Comment by akshayjshah** [comment]
Glad that we discovered this race now! I agree with Peter that closures feel like an unusual way to maintain private state in Go.

Would this feel more idiomatic if had something like this?

```
type Strategy interface {
  Backoff(attempts uint) time.Duration
}

// optionally, we could also have
type BackoffFunc func(uint) time.Duration
func (f BackoffFunc) Backoff(n uint) time.Duration { return f(n) }

func NewExponential(opts ...ExponentialOption) (Strategy, error) { ... }
```

File: internal/backoff/exponential.go [hunk 1]
```diff
@@ -35,8 +35,7 @@ type ExponentialOption func(*exponentialOptions)
 // exponentialOptions are the configuration options for an exponential backoff
 type exponentialOptions struct {
 	base, min, max time.Duration
```
Comment (akshayjshah) [inline internal/backoff/exponential.go]: Under what circumstances would I want a non-zero minimum? As long as we're jittering, isn't zero safe for the server and fastest for the client?

Comment (akshayjshah) [inline internal/backoff/exponential.go]: Also...consider whether we should choose a new name for `base`. (Perhaps `initial`?)

In the context of exponentiation, I expected `base` to be `M` in `M ^ n`. That made me immediately suspicious of the `time.Duration` type.

File: internal/backoff/exponential.go [hunk 2]
```diff
@@ -55,12 +54,45 @@ func (e exponentialOptions) validate() (err error) {
...
+}
+
+// DefaultExponential is an exponential backoff.Strategy configured
+// with a 100ms minimum delay, a 100ms base for a jittered exponential backoff
+// curve, and a maximum time to recovery of one minute.
+//
+// Exponential strategies are not thread safe.
+// Use the Isolate() method to obtain an independent backoff generator with the
+// same options and a referentially independent random number generator.
+var DefaultExponential = (&exponentialStrategy{
```
Comment (abhinav) [inline internal/backoff/exponential.go]: If this is a `backoff.Strategy`, annotate the var with the type. Otherewise the type of `DefaultExponential` will be `func(uint) Duration`.

    var DefaultExponential Strategy = ...

File: internal/backoff/exponential.go (hunk 1 above)
Comment (willhug) [inline internal/backoff/exponential.go]: whoops, good point.

File: internal/backoff/exponential.go [hunk 3]
```diff
@@ -84,24 +116,28 @@ func MinBackoff(t time.Duration) ExponentialOption {
...
-type Exponential struct {
-	opts exponentialOptions
-}
-
-// NewExponential returns a new Exponential Backoff Strategy.
-func NewExponential(opts ...ExponentialOption) (*Exponential, error) {
+//
+// Backoff functions are lockless and referentially independent, but not
+// thread-safe.
+func NewExponential(opts ...ExponentialOption) (func() func(uint) time.Duration, error) {
```
Comment (abhinav) [inline internal/backoff/exponential.go]: Shouldn't this return a Strategy?

**Comment by kriskowal** [comment]
Is everyone in favor of adding a backoff.Strategy interface to `yarpc/api/backoff`? It’ll have to be moved out of internal since it’s an argument to public ConnBackoff transport options.

**Comment by akshayjshah** [comment]
I'd prefer to add a strategy interface instead of committing to the function type in a public API. I can be convinced, though :)

**Comment by kriskowal** [comment]
I’ve posted changes to address feedback and incidentally expand the scope of the change.

I’ve created an api/backoff with Strategy and Backoff interfaces. These are threaded through later changes to the options of TChannel and HTTP transport ConnBackoff constructors.

I’ve also changed the implementation to use only (first, max) parameters, removing (min) and renaming (base). These changes I’ve threaded through the config change following this one.

I’ve renamed the "Short" strategy and confined its type to the Strategy interface since it isn’t necessarily implemented as a degenerate ExponentialStrategy. It’s effectively a Noop type and could be implemented as a singleton in a future pass, but I’ve left it as is for expedience.

I think that I might send a note back in time suggesting that all API packages should have an "api" suffix like "peerapi" and "backoffapi", since it’s become pretty common to have to import both the API and implementation packages.

File: internal/backoff/exponential.go [hunk 4]
```diff
@@ -34,37 +35,47 @@ type ExponentialOption func(*exponentialOptions)
 
 // exponentialOptions are the configuration options for an exponential backoff
 type exponentialOptions struct {
-	base, min, max time.Duration
-	rand           *rand.Rand
-	minMaxDiff     int64
+	first, max time.Duration
```
Comment (peter-edge) [inline internal/backoff/exponential.go]: nit: I don't like this style of declaring multiple variables with the same type in the same line, and we don't do it in other places in the codebase I don't think

File: internal/backoff/exponential.go [hunk 5]
```diff
@@ -34,37 +35,47 @@ type ExponentialOption func(*exponentialOptions)
...
 }
 
 func (e exponentialOptions) validate() (err error) {
-	if e.base <= 0 {
-		err = multierr.Append(err, errors.New("invalid base for exponential backoff, need greater than zero"))
-	}
-	if e.min < 0 {
-		err = multierr.Append(err, errors.New("invalid min for exponential backoff, need greater than or equal to zero"))
+	if e.first <= 0 {
+		err = multierr.Append(err, errors.New("invalid first duration for exponential backoff, need greater than zero"))
```
Comment (peter-edge) [inline internal/backoff/exponential.go]: nit (but not really part of this PR): I like most errors created with `errors.New` to be global constants.

File: internal/backoff/exponential.go [hunk 6]
```diff
@@ -34,37 +35,47 @@ type ExponentialOption func(*exponentialOptions)
...
 
+func newRand() *rand.Rand {
+	return rand.New(rand.NewSource(time.Now().UnixNano()))
+}
+
 var defaultExponentialOpts = exponentialOptions{
-	base: time.Millisecond,
-	max:  time.Hour, // :shrug:
-	rand: rand.New(rand.NewSource(time.Now().UnixNano())),
+	first:   100 * time.Millisecond,
```
Comment (peter-edge) [inline internal/backoff/exponential.go]: Can we make `defaultExponentialOpts` a public variable so this can be seen in documentation?  We should also talk about all the arbitrary default durations around the codebase.

File: internal/backoff/exponential.go [hunk 7]
```diff
@@ -26,6 +26,7 @@ import (
 	"time"
 
 	"go.uber.org/multierr"
+	"go.uber.org/yarpc/api/backoff"
```
Comment (willhug) [inline internal/backoff/exponential.go]: ordering?

File: internal/backoff/exponential_test.go [hunk 8]
```diff
@@ -39,57 +46,16 @@ func TestExponential(t *testing.T) {
...
-		{
-			msg:      "invalid min",
-			giveBase: time.Duration(1000),
-			giveMax:  time.Duration(0),
-			giveMin:  time.Duration(-100),
-			wantErrors: []string{
-				"invalid min for exponential backoff, need greater than or equal to zero",
-			},
-		},
-		{
```
Comment (willhug) [inline internal/backoff/exponential_test.go]: we can keep the invalid max test
